Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -4875,27 +4878,51 @@ - (void)reloadViewItems
                 [MessageRecipientStatusUtils recipientStatusWithOutgoingMessage:outgoingMessage
                                                                   referenceView:self.view];
 
-            if (outgoingMessage.messageState == TSOutgoingMessageStateFailed) {
-                // always show "failed to send" status
-                shouldHideRecipientStatus = NO;
-            } else {
-                shouldHideRecipientStatus
-                    = (interactionType == lastInteractionType && receiptStatus == lastReceiptStatus);
-            }
+            // Always show "failed to send" status.
+            shouldHideFooter = (interactionType == lastInteractionType && receiptStatus == lastReceiptStatus
+                && outgoingMessage.messageState != TSOutgoingMessageStateFailed);
 
             lastReceiptStatus = receiptStatus;
         } else if (interactionType == OWSInteractionType_IncomingMessage) {
             TSIncomingMessage *incomingMessage = (TSIncomingMessage *)viewItem.interaction;
             NSString *incomingSenderId = incomingMessage.authorId;
             OWSAssert(incomingSenderId.length > 0);
-            shouldHideAvatar = (interactionType == lastInteractionType &&
+            BOOL isCollapsed = (interactionType == lastInteractionType &&
                 [NSObject isNullableObject:lastIncomingSenderId equalTo:incomingSenderId]);
             lastIncomingSenderId = incomingSenderId;
+
+            shouldShowSenderAvatar = viewItem.isGroupThread && !isCollapsed;
+        }
+        lastInteractionType = interactionType;
+
+        viewItem.shouldShowSenderAvatar = shouldShowSenderAvatar;
+        viewItem.shouldHideFooter = shouldHideFooter;
+    }
+
+    // Iterate again in forward order.
```

Comment:
PERF TODO: not sure if this is a hotspot, but it seems like we could combine these iterations by keeping a reference to the previousViewItem.

It's OK for this PR though.